clippy_lints/src/ptr.rs

```diff
@@ -189,53 +190,50 @@ impl<'tcx> LateLintPass<'tcx> for Ptr {
-        (&paths::PTR_READ, 0, 1),
-        (&paths::PTR_READ_UNALIGNED, 0, 1),
-        (&paths::PTR_READ_VOLATILE, 0, 1),
-        (&paths::PTR_REPLACE, 0, 2),
-        (&paths::PTR_SLICE_FROM_RAW_PARTS, 0, 2),
-        (&paths::PTR_SLICE_FROM_RAW_PARTS_MUT, 0, 2),
-        (&paths::PTR_SWAP, 0, 2),
-        (&paths::PTR_SWAP, 1, 2),
-        (&paths::PTR_SWAP_NONOVERLAPPING, 0, 3),
-        (&paths::PTR_SWAP_NONOVERLAPPING, 1, 3),
-        (&paths::PTR_WRITE, 0, 2),
-        (&paths::PTR_WRITE_UNALIGNED, 0, 2),
-        (&paths::PTR_WRITE_VOLATILE, 0, 2),
-        (&paths::PTR_WRITE_BYTES, 0, 3),
+    // (fn_path, arg_indices) - `arg_indices` are the `arg` positions where null would cause U.B.
+    const INVALID_NULL_PTR_USAGE_TABLE: [(&[&str], &[usize]); 16] = [
+        (&paths::SLICE_FROM_RAW_PARTS, &[0]),
+        (&paths::SLICE_FROM_RAW_PARTS_MUT, &[0]),
+        (&paths::PTR_COPY, &[0, 1]),
+        (&paths::PTR_COPY_NONOVERLAPPING, &[0, 1]),
+        (&paths::PTR_READ, &[0]),
+        (&paths::PTR_READ_UNALIGNED, &[0]),
+        (&paths::PTR_READ_VOLATILE, &[0]),
+        (&paths::PTR_REPLACE, &[0]),
+        (&paths::PTR_SLICE_FROM_RAW_PARTS, &[0]),
+        (&paths::PTR_SLICE_FROM_RAW_PARTS_MUT, &[0]),
+        (&paths::PTR_SWAP, &[0, 1]),
+        (&paths::PTR_SWAP_NONOVERLAPPING, &[0, 1]),
+        (&paths::PTR_WRITE, &[0]),
+        (&paths::PTR_WRITE_UNALIGNED, &[0]),
+        (&paths::PTR_WRITE_VOLATILE, &[0]),
+        (&paths::PTR_WRITE_BYTES, &[0]),
     ];
 
-    for (fn_path, arg_idx, num_args) in &INVALID_NULL_PTR_USAGE_TABLE {
-        if_chain! {
-            if let ExprKind::Call(ref fun, ref args) = expr.kind;
-            if args.len() == *num_args;
-            if let ExprKind::Path(ref qpath) = fun.kind;
-            if let Some(arg) = args.get(*arg_idx);
-            if let Some(fun_def_id) = cx.qpath_res(qpath, fun.hir_id).opt_def_id();
-            if match_def_path(cx, fun_def_id, fn_path);
-            if is_null_path(cx, arg);
-            then {
-                span_lint_and_sugg(
-                    cx,
-                    INVALID_NULL_PTR_USAGE,
-                    arg.span,
-                    "pointer must be non-null",
-                    "change this to",
-                    "core::ptr::NonNull::dangling().as_ptr()".to_string(),
-                    Applicability::MachineApplicable,
-                );
-                return;
+    if_chain! {
+        if let ExprKind::Call(ref fun, ref args) = expr.kind;
+        if let ExprKind::Path(ref qpath) = fun.kind;
+        if let Some(fun_def_id) = cx.qpath_res(qpath, fun.hir_id).opt_def_id();
+        let fun_def_path = cx.get_def_path(fun_def_id).into_iter().map(Symbol::to_ident_string).collect::<Vec<_>>();
+        if let Some(arg) = INVALID_NULL_PTR_USAGE_TABLE.iter().find_map(|(fn_path, arg_indices)| {
+            if fn_path == &fun_def_path {
+                arg_indices.iter().find_map(|arg_idx| {
```

Comment:
It would be better to lint multiple args rather than stopping at the first one (even though that's pretty unlikely).